### DISCUSSION TITLE
Cache unexpected behavior with react-native on Android

### DISCUSSION THREAD

**Discussion by devpolo** [TanStack/query#2694] [Q&A]
## Hi every one 👋

Hope you are all doing well 
It's been days since I’m facing a react-query problem. 

I am using react-query in a react-native app.
In iOS, everything is working as expected. 
On the other hand, on Android, the cache is never invalidate, even after the default 5 minutes (by the way, I didn’t modify the default options staleTime and cacheTime because they meet my needs). 

For exemple, if I change a product title in my backend, it takes more than 24 hours to get update on my Android app (and 5 minutes on iOS). 

Is there any one who can help me understand what is happening ? 

react-native: 0.63.1
react-query: 3.9.8

**Comment by TkDodo** [comment]
not sure I fully understand, but my experience with react-native is very limited. 

> In iOS, everything is working as expected.

what is expected please? Do you have an example?

> On the other hand, on Android, the cache is never invalidate, even after the default 5 minutes

cacheTime of 5 minutes has nothing to do with query invalidation. When a query is _unused_ (=not rendered on the screen), it will be removed from the cache (garbage collected) after 5 minutes. But doesn't refetch it. I'm going into more details here: https://tkdodo.eu/blog/practical-react-query#the-defaults-explained

> For exemple, if I change a product title in my backend, it takes more than 24 hours to get update on my Android app (and 5 minutes on iOS).

How so?  With a staleTime of 0, whenever the component mounts or you e.g. focus the window (not sure how that works in native tbh), a refetch should happen.

**Comment by SamiChab** [comment]
Hello @TkDodo 

First of all, thank you for your answer. I'm actually working with @devpolo and I will follow up this question.

> what is expected please? Do you have an example?

We expect to get the fresh data from the backend when launching the app

> How so? With a staleTime of 0, whenever the component mounts or you e.g. focus the window (not sure how that works in native tbh), a refetch should happen

Yes, it should happen, but it doesn't... No way to get fresh data with `useQuery` and `useInfiniteQuery` hooks. I've tried to manually set `cacheTime: 1000` or to set `refetchOnMount: 'always'`, but I always get data from cache. 

react-native: 0.63.1
react-query: 3.9.8

**Reply by TkDodo** [reply]
> We expect to get the fresh data from the backend when launching the app

why would you not get fresh data when launching the app? Unless you persist the queryClient somewhere, if the app is opened, there is no cache, so data must come from the backend.

> react-query: 3.9.8

the latest version is 3.31.0, so please consider updating. other than that, without getting a reproducible example or at least seeing some code, it's impossible to say.

**Reply by SamiChab** [reply]
@TkDodo thanks for your answer.

> the latest version is 3.31.0, so please consider updating. other than that, without getting a reproducible example or at least seeing some code, it's impossible to say.

I just installed the latest version of react-query: `3.33.0`

> why would you not get fresh data when launching the app? Unless you persist the queryClient somewhere, if the app is opened, there is no cache, so data must come from the backend.


That's what I try to figure out. Even with the new version of the app, I get old data from backend.

Here's a video where I show the problem. I added a `.` after the sentence `Emprunter un chemin intérieur` in the backend, but it doesn't get fetched by react-query whereas it works in Postman...

https://user-images.githubusercontent.com/28505397/142254857-1bffbc84-e79f-4814-8c69-0a61d4802fbd.mov

My `react-query` client:

```
const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      refetchOnMount: 'always',
      staleTime: 0,
      cacheTime: 0,
      queryFn: defaultQueryFn,
      onError,
    },
    mutations: {
      onError,
    },
  },
})
```

(I set `refetchOnMount`, `cacheTime` and `staleTime` for purpose of testing)

My hook:
```
export const useQueryPrograms = (
  params: IQueryProgramsParams = {},
): QueryObserverResult<IProgram[]> => {
  return useQuery(['/programs', params])
}
```

**Reply by TkDodo** [reply]
Is there a way to see the actual network request that’s going out? Could be a different cache layer that returns cached data…

**Reply by SamiChab** [reply]
Yes, here it is

<img width="545" alt="Capture d’écran 2021-11-17 à 21 09 06" src="https://user-images.githubusercontent.com/28505397/142274699-10205e1a-015c-477c-9bbe-d8eb8e7a4314.png">

**Reply by TkDodo** [reply]
`max-age=14400` indicates that the browser can cache the resource for 4 hours, which seems to be the default for cloudfare. I'm guessing that the request is going out but being served from the browser cache.

**Reply by SamiChab** [reply]
@TkDodo interesting. Have you an idea of what I can do to change this duration ?

**Reply by TkDodo** [reply]
https://community.cloudflare.com/t/new-cloudflares-rules-on-cache-control/69770

**Reply by SamiChab** [reply]
@TkDodo thank you so much !! It works now. The backend developer set following header parameters: `max-age=5, public, s-maxage=30`
